### ISSUE TITLE
[env_dpdk_post_init] pci_device_rte_dev_event heap-use-after-free after pci_unplug

### ISSUE THREAD

**Issue by mgerdts** [spdk/spdk#2983]
# CI Intermittent Failure

## Link to the failed CI build

https://ci.spdk.io/results/autotest-per-patch/builds/102045/archive/BlobFS-autotest/build.log

## Execution failed at

```
00:04:04.571  ************************************
00:04:04.571  START TEST env_dpdk_post_init
00:04:04.571  ************************************
00:04:04.571   17:46:35	-- common/autotest_common.sh@1075 -- # xtrace_enable
00:04:04.571   17:46:35	-- common/autotest_common.sh@1076 -- # /var/jenkins/workspace/BlobFS-autotest/spdk/test/env/env_dpdk_post_init/env_dpdk_post_init -c 0x1 --base-virtaddr=0x200000000000
00:04:04.571  EAL: Detected CPU lcores: 96
00:04:04.571  EAL: Detected NUMA nodes: 2
00:04:04.571  EAL: Detected shared linkage of DPDK
00:04:04.571  EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
00:04:04.571  EAL: Selected IOVA mode 'VA'
00:04:04.571  EAL: No free 2048 kB hugepages reported on node 1
00:04:04.571  EAL: VFIO support initialized
00:04:04.571  TELEMETRY: No legacy callbacks, legacy socket not created
00:04:04.831  EAL: Using IOMMU type 1 (Type 1)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:00:04.0 (socket 0)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:00:04.1 (socket 0)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:00:04.2 (socket 0)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:00:04.3 (socket 0)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:00:04.4 (socket 0)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:00:04.5 (socket 0)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:00:04.6 (socket 0)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:00:04.7 (socket 0)
00:04:04.831  EAL: Ignore mapping IO port bar(1)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:80:04.0 (socket 1)
00:04:04.831  EAL: Ignore mapping IO port bar(1)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:80:04.1 (socket 1)
00:04:04.831  EAL: Ignore mapping IO port bar(1)
00:04:04.831  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:80:04.2 (socket 1)
00:04:04.831  EAL: Ignore mapping IO port bar(1)
00:04:05.090  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:80:04.3 (socket 1)
00:04:05.090  EAL: Ignore mapping IO port bar(1)
00:04:05.090  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:80:04.4 (socket 1)
00:04:05.090  EAL: Ignore mapping IO port bar(1)
00:04:05.090  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:80:04.5 (socket 1)
00:04:05.090  EAL: Ignore mapping IO port bar(1)
00:04:05.090  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:80:04.6 (socket 1)
00:04:05.090  EAL: Ignore mapping IO port bar(1)
00:04:05.090  EAL: Probe PCI driver: spdk_ioat (8086:2021) device: 0000:80:04.7 (socket 1)
00:04:05.661  EAL: Probe PCI driver: spdk_nvme (8086:a54) device: 0000:d9:00.0 (socket 1)
00:04:09.867  EAL: Releasing PCI mapped resource for 0000:00:04.7
00:04:09.867  EAL: Calling pci_unmap_resource for 0000:00:04.7 at 0x20200101c000
00:04:09.867  EAL: Releasing PCI mapped resource for 0000:00:04.6
00:04:09.867  EAL: Calling pci_unmap_resource for 0000:00:04.6 at 0x202001018000
00:04:09.867  EAL: Releasing PCI mapped resource for 0000:00:04.5
00:04:09.867  EAL: Calling pci_unmap_resource for 0000:00:04.5 at 0x202001014000
00:04:09.867  EAL: Releasing PCI mapped resource for 0000:00:04.4
00:04:09.867  EAL: Calling pci_unmap_resource for 0000:00:04.4 at 0x202001010000
00:04:09.867  EAL: Releasing PCI mapped resource for 0000:00:04.3
00:04:09.867  EAL: Calling pci_unmap_resource for 0000:00:04.3 at 0x20200100c000
00:04:09.867  EAL: Releasing PCI mapped resource for 0000:00:04.2
00:04:09.867  EAL: Calling pci_unmap_resource for 0000:00:04.2 at 0x202001008000
00:04:09.867  EAL: Releasing PCI mapped resource for 0000:00:04.1
00:04:09.867  EAL: Calling pci_unmap_resource for 0000:00:04.1 at 0x202001004000
00:04:09.867  EAL: Releasing PCI mapped resource for 0000:00:04.0
00:04:09.868  EAL: Calling pci_unmap_resource for 0000:00:04.0 at 0x202001000000
00:04:09.868  =================================================================
00:04:09.868  ==1870287==ERROR: AddressSanitizer: heap-use-after-free on address 0x6120000281d0 at pc 0x7f1e37f7f276 bp 0x7f1e33ffd8b0 sp 0x7f1e33ffd060
00:04:09.868  READ of size 6 at 0x6120000281d0 thread T1
00:04:09.868      #0 0x7f1e37f7f275 in __interceptor_strcmp.part.0 (/usr/lib64/libasan.so.8+0xaa275)
00:04:09.868      #1 0x7f1e372a50c9 in pci_device_rte_dev_event /var/jenkins/workspace/BlobFS-autotest/spdk/lib/env_dpdk/pci.c:236
00:04:09.868      #2 0x7f1e37082e20 in rte_dev_event_callback_process ../lib/eal/common/eal_common_dev.c:604
00:04:09.868      #3 0x7f1e371d799c in pci_hot_unplug_handler ../drivers/bus/pci/pci_common.c:622
00:04:09.868      #4 0x7f1e371dc6a4 in pci_vfio_req_handler ../drivers/bus/pci/linux/pci_vfio.c:365
00:04:09.868      #5 0x7f1e370b6a10 in eal_intr_process_interrupts ../lib/eal/linux/eal_interrupts.c:1017
00:04:09.868      #6 0x7f1e370b6d18 in eal_intr_handle_interrupts ../lib/eal/linux/eal_interrupts.c:1091
00:04:09.868      #7 0x7f1e370b6eff in eal_intr_thread_main ../lib/eal/linux/eal_interrupts.c:1163
00:04:09.868      #8 0x7f1e37097481 in ctrl_thread_init ../lib/eal/common/eal_common_thread.c:267
00:04:09.868      #9 0x7f1e365c212c in start_thread (/usr/lib64/libc.so.6+0x8b12c)
00:04:09.868      #10 0x7f1e36643bbf in __clone3 (/usr/lib64/libc.so.6+0x10cbbf)
00:04:09.868  
00:04:09.868  0x6120000281d0 is located 272 bytes inside of 312-byte region [0x6120000280c0,0x6120000281f8)
00:04:09.868  freed by thread T1 here:
00:04:09.868      #0 0x7f1e37f8e388 in __interceptor_free.part.0 (/usr/lib64/libasan.so.8+0xb9388)
00:04:09.868      #1 0x7f1e371d68f8 in pci_free ../drivers/bus/pci/pci_common.c:129
00:04:09.868      #2 0x7f1e371d7af8 in pci_unplug ../drivers/bus/pci/pci_common.c:682
00:04:09.868      #3 0x7f1e37082008 in local_dev_remove ../lib/eal/common/eal_common_dev.c:359
00:04:09.868      #4 0x7f1e3708222d in rte_dev_remove ../lib/eal/common/eal_common_dev.c:440
00:04:09.868      #5 0x7f1e37081fa0 in rte_eal_hotplug_remove ../lib/eal/common/eal_common_dev.c:344
00:04:09.868      #6 0x7f1e372a42bd in remove_rte_dev /var/jenkins/workspace/BlobFS-autotest/spdk/lib/env_dpdk/pci.c:109
00:04:09.868      #7 0x7f1e372a43d1 in detach_rte_cb /var/jenkins/workspace/BlobFS-autotest/spdk/lib/env_dpdk/pci.c:116
00:04:09.868      #8 0x7f1e370b15e2 in eal_alarm_callback ../lib/eal/linux/eal_alarm.c:104
00:04:09.868      #9 0x7f1e370b6a10 in eal_intr_process_interrupts ../lib/eal/linux/eal_interrupts.c:1017
00:04:09.868      #10 0x7f1e370b6d18 in eal_intr_handle_interrupts ../lib/eal/linux/eal_interrupts.c:1091
00:04:09.868      #11 0x7f1e370b6eff in eal_intr_thread_main ../lib/eal/linux/eal_interrupts.c:1163
00:04:09.868      #12 0x7f1e37097481 in ctrl_thread_init ../lib/eal/common/eal_common_thread.c:267
00:04:09.868      #13 0x7f1e365c212c in start_thread (/usr/lib64/libc.so.6+0x8b12c)
00:04:09.868  
00:04:09.868  previously allocated by thread T0 here:
00:04:09.868      #0 0x7f1e37f8f6af in __interceptor_malloc (/usr/lib64/libasan.so.8+0xba6af)
00:04:09.868      #1 0x7f1e371d94a4 in pci_scan_one ../drivers/bus/pci/linux/pci.c:218
00:04:09.868      #2 0x7f1e371d9eb2 in rte_pci_scan ../drivers/bus/pci/linux/pci.c:471
00:04:09.868      #3 0x7f1e3708094c in rte_bus_scan ../lib/eal/common/eal_common_bus.c:56
00:04:09.868      #4 0x7f1e370b0847 in rte_eal_init ../lib/eal/linux/eal.c:1065
00:04:09.868      #5 0x402ac5 in main /var/jenkins/workspace/BlobFS-autotest/spdk/test/env/env_dpdk_post_init/env_dpdk_post_init.c:73
00:04:09.868      #6 0x7f1e3655e50f in __libc_start_call_main (/usr/lib64/libc.so.6+0x2750f)
00:04:09.868  
00:04:09.868  Thread T1 created by T0 here:
00:04:09.868      #0 0x7f1e37f203e6 in __interceptor_pthread_create (/usr/lib64/libasan.so.8+0x4b3e6)
00:04:09.868      #1 0x7f1e37097509 in rte_ctrl_thread_create ../lib/eal/common/eal_common_thread.c:288
00:04:09.868      #2 0x7f1e370b6f98 in rte_eal_intr_init ../lib/eal/linux/eal_interrupts.c:1191
00:04:09.868      #3 0x7f1e370b0789 in rte_eal_init ../lib/eal/linux/eal.c:1037
00:04:09.868      #4 0x402ac5 in main /var/jenkins/workspace/BlobFS-autotest/spdk/test/env/env_dpdk_post_init/env_dpdk_post_init.c:73
00:04:09.868      #5 0x7f1e3655e50f in __libc_start_call_main (/usr/lib64/libc.so.6+0x2750f)
00:04:09.868  
00:04:09.868  SUMMARY: AddressSanitizer: heap-use-after-free (/usr/lib64/libasan.so.8+0xaa275) in __interceptor_strcmp.part.0
00:04:09.868  Shadow bytes around the buggy address:
00:04:09.868    0x0c247fffcfe0: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
00:04:09.868    0x0c247fffcff0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
00:04:09.868    0x0c247fffd000: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fa
00:04:09.868    0x0c247fffd010: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
00:04:09.868    0x0c247fffd020: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
00:04:09.868  =>0x0c247fffd030: fd fd fd fd fd fd fd fd fd fd[fd]fd fd fd fd fa
00:04:09.868    0x0c247fffd040: fa fa fa fa fa fa fa fa 00 00 00 00 00 00 00 00
00:04:09.868    0x0c247fffd050: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00:04:09.868    0x0c247fffd060: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fa
00:04:09.868    0x0c247fffd070: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
00:04:09.868    0x0c247fffd080: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
00:04:09.868  Shadow byte legend (one shadow byte represents 8 application bytes):
00:04:09.868    Addressable:           00
00:04:09.868    Partially addressable: 01 02 03 04 05 06 07 
00:04:09.868    Heap left redzone:       fa
00:04:09.868    Freed heap region:       fd
00:04:09.868    Stack left redzone:      f1
00:04:09.868    Stack mid redzone:       f2
00:04:09.868    Stack right redzone:     f3
00:04:09.868    Stack after return:      f5
00:04:09.868    Stack use after scope:   f8
00:04:09.868    Global redzone:          f9
00:04:09.868    Global init order:       f6
00:04:09.868    Poisoned by user:        f7
00:04:09.868    Container overflow:      fc
00:04:09.868    Array cookie:            ac
00:04:09.868    Intra object redzone:    bb
00:04:09.868    ASan internal:           fe
00:04:09.868    Left alloca redzone:     ca
00:04:09.868    Right alloca redzone:    cb
00:04:09.868  ==1870287==ABORTING
00:04:09.868    17:46:40	-- common/autotest_common.sh@1076 -- # trap - ERR
00:04:09.868    17:46:40	-- common/autotest_common.sh@1076 -- # print_backtrace
00:04:09.868    17:46:40	-- common/autotest_common.sh@1104 -- # [[ ehxBET =~ e ]]
00:04:09.868    17:46:40	-- common/autotest_common.sh@1106 -- # args=('--base-virtaddr=0x200000000000' '0x1' '-c' '/var/jenkins/workspace/BlobFS-autotest/spdk/test/env/env_dpdk_post_init/env_dpdk_post_init' 'env_dpdk_post_init')
00:04:09.868    17:46:40	-- common/autotest_common.sh@1106 -- # local args
00:04:09.868    17:46:40	-- common/autotest_common.sh@1108 -- # xtrace_disable
00:04:09.868    17:46:40	-- common/autotest_common.sh@20 -- # '[' no '!=' yes ']'
00:04:09.868    17:46:40	-- common/autotest_common.sh@21 -- # PREV_BASH_OPTS=ehxBET
00:04:09.868    17:46:40	-- common/autotest_common.sh@22 -- # [[ ehxBET == *\x* ]]
00:04:09.869    17:46:40	-- common/autotest_common.sh@23 -- # XTRACE_DISABLED=yes
00:04:09.869    17:46:40	-- common/autotest_common.sh@25 -- # set +x
00:04:09.869  ========== Backtrace start: ==========
00:04:09.869  
00:04:09.869  in /var/jenkins/workspace/BlobFS-autotest/spdk/test/common/autotest_common.sh:1076 -> run_test(["env_dpdk_post_init"],["/var/jenkins/workspace/BlobFS-autotest/spdk/test/env/env_dpdk_post_init/env_dpdk_post_init"],["-c"],["0x1"],["--base-virtaddr=0x200000000000"])
00:04:09.869       ...
00:04:09.869     1071		timing_enter $test_name
00:04:09.869     1072		echo "************************************"
00:04:09.869     1073		echo "START TEST $test_name"
00:04:09.869     1074		echo "************************************"
00:04:09.869     1075		xtrace_restore
00:04:09.869     1076		time "$@"
00:04:09.869     1077		xtrace_disable
00:04:09.869     1078		echo "************************************"
00:04:09.869     1079		echo "END TEST $test_name"
00:04:09.869     1080		echo "************************************"
00:04:09.869     1081		timing_exit $test_name
00:04:09.869       ...
00:04:09.869  in /var/jenkins/workspace/BlobFS-autotest/spdk/test/env/env.sh:24 -> main([])
00:04:09.869       ...
00:04:09.869     19  		# We speed up the process by specifying an address that's not
00:04:09.869     20  		# supposed to be reserved by ASAN. Regular SPDK applications do
00:04:09.869     21  		# this implicitly.
00:04:09.869     22  		argv+="--base-virtaddr=0x200000000000"
00:04:09.869     23  	fi
00:04:09.869  => 24  	run_test "env_dpdk_post_init" $testdir/env_dpdk_post_init/env_dpdk_post_init $argv
00:04:09.869     25  	
00:04:09.869     26  	if [ $(uname) = Linux ]; then
00:04:09.869     27  		# This tests the --match-allocations DPDK parameter which is only
00:04:09.869     28  		# supported on Linux
00:04:09.869     29  		run_test "env_mem_callbacks" $testdir/mem_callbacks/mem_callbacks
00:04:09.869       ...
```

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mgerdts. log: https://ci.spdk.io/public_build/autotest-per-patch_102045.html

**Comment by jimharris** [comment]
> Some similarities with issue #2953, but not enough to say they are the same issue.

I don't think these are related now.

I tried to reproduce this in Jenkins CI, running this test a 500 times in a row across 4 different job instances.  I wasn't able to repro.

I've added https://review.spdk.io/gerrit/c/spdk/spdk/+/17631 which I'd like to get merged to help narrow down the problem.  It seems like after we detach in this test app, we trigger DPDK to free the associated PCI device, but then there's another callback within DPDK via vfio which triggers an PCI remove callback up to SPDK on this same device.

I think the fix here is likely to just switch the order of the two clauses on line 243 - only do the strcmp if the pending_removal flag is not set.  But the ERRLOGs here will first help confirm we're on the right track.
